README.md
![patteRNA](patterna.png)

# *patteRNA*

Rapid mining of RNA secondary structure motifs from structure profiling data.

[![License](https://img.shields.io/badge/License-BSD%202--Clause-orange.svg)](https://opensource.org/licenses/BSD-2-Clause)

## What Is It?

*patteRNA* is an unsupervised pattern recognition algorithm that rapidly mines RNA structure motifs from structure profiling (SP) data. 

It features a discretized observation model, hidden Markov model (DOM-HMM) of reactivity that enables automated and calibrated processing of SP data without a dependence on reference structures. It is compatible with most current probing techniques (e.g., SHAPE, DMS, PARS) and can help analyze datasets of any size, from small scale experiments to transcriptome-wide assays. It scales well to millions or billions of nucleotides.

The training and scoring implementations are parallelized, so the algorithm can benefit greatly when deployed in a high CPU-count environment.



## Getting Started

These instructions will set up patteRNA as a command line tool.

### Prerequisites

**Python 3.7 or newer**. To check the current version on your system, run `python3 -V`. If your version is anterior to 3.7, install version 3.7 or later. We recommend the use of virtual environments with [venv](https://docs.python.org/3/library/venv.html). You also need the latest versions of `pip` and `setuptools`, which can be installed by typing: `python -m pip install -U pip setuptools`.

**ViennaRNA Python interface**. [ViennaRNA](https://www.tbi.univie.ac.at/RNA/) is a C library of RNA folding routines. In order for patteRNA to most accurately identify motifs, the [ViennaRNA Python interface](https://www.tbi.univie.ac.at/RNA/documentation.html#interfaces) must be installed and configured for your Python environment. You should be able to run `python -c import RNA` without errors. If the ViennaRNA interface is not detected, patteRNA can still mine motifs but will be slightly less precise. Use the flag `--no-vienna` to avoid warnings.

### Installation

Installation is done directly from source. For that, clone this repository using the commands:

```
git clone https://github.com/AviranLab/patteRNA.git
cd patteRNA
```

To install the Python module such that it can be executed from the command line, run the setup script.

```
python setup.py install
```

You can also specify a local installation using the commands:

```
python setup.py install --user
echo 'export PATH="$PATH:~/.local/bin"' >> ~/.bashrc; source ~/.bashrc
```

***Note for macOS Big Sur users***: Due to an issue, you must use `pip` to run the installation. Be sure to update pip and setuptools before attempting the installation (`python -m pip install -U pip setuptools`). Use the commands:
```
python -m pip install .
```
or
```
python -m pip install . --user
```

***Note for Apple Silicon users (M1)***: Some dependencies are not yet pre-built for Apple Silicon, so you must install them manually if using a native Python executable (i.e., non-`x86_64`). Specifically, to install `numpy`, `scipy`, and `sklearn` properly, follow [these instructions](https://github.com/scipy/scipy/issues/13409) before installing with `pip`.

### Running a test

To make sure patteRNA is properly installed, run the following command:

```
patteRNA --version
```

This should output the current version of patteRNA. You can now do a test by entering the following command:

```
patteRNA sample_data/weeks_set.shape sample_output -f sample_data/weeks_set.fa --motif "((((...))))" -v
```

This will run patteRNA in verbose mode (`-v`) and create an output directory `sample_output` in the current folder.



## Usage

### General usage and available options

```
patteRNA <probing> <output> <OPTIONS>
```

All available options are accessible via `patteRNA -h` as listed below. Recommendations (when applicable) are given in the option caption. Note that switches (i.e. boolean options that do not need arguments), have defaults set to `False` and are set to `True` if provided.

```
usage: patteRNA [-h] [--version] [-v] [-f fasta] [--reference reference] [-l]
                [--no-vienna] [--GMM] [-k kernels] [--KL-div KL-div] [-e eps]
                [-i iter] [-t tasks] [--model model] [--motif motif]
                [--path path] [--hairpins] [--posteriors] [--viterbi] [--HDSL]
                [--SPP] [--nan] [--print-nan] [--no-prompt]
                [--min-cscores min] [--no-cscores] [--batch-size size]
                [-c length]
                probing output

Rapid mining of RNA secondary structure motifs from profiling data.

positional arguments:
  probing               FASTA-like file of probing data
  output                Output directory

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  -v, --verbose         Print detailed progress logs (default: False)
  -f fasta, --fasta fasta
                        FASTA file of RNA sequences (default: None)
  --reference reference
                        FASTA-like file of reference RNA secondary structures
                        in dot-bracket notation (default: None)
  -l, --log             Log transform input data (default: False)
  --no-vienna           Do not attempt to use ViennaRNA libraries. Turns off
                        LBC scoring classifier. (default: False)
  --GMM                 Train a Gaussian Mixture Model (GMM) during training
                        instead of a Discretized ObservationModel (DOM)
                        (default: False)
  -k kernels            Number of kernels per pairing state to use in the
                        emission model. By default, k is determined
                        automatically using Bayesian Information Criteria.
                        Increasing k manually can more precisely fit=the data,
                        but could result in overfitting. Fitted data should
                        always be visually inspected after training to gauge
                        if the model is adequate. (default: -1)
  --KL-div KL-div       Minimum Kullback–Leibler divergence criterion for
                        building the training set. The KL divergence measures
                        the difference in information content between the full
                        dataset and the training set. The smaller the value,
                        the more representative the training set will be with
                        respect to the full dataset. However, this will
                        produce a larger training set and increase both
                        runtime and RAM consumption during training. (default:
                        0.001)
  -e eps, --epsilon eps
                        Convergence criterion (default: 0.01)
  -i iter, --maxiter iter
                        Maximum number of training iterations (default: 250)
  -t tasks, --n-tasks tasks
                        Number of parallel processes. By default all available
                        CPUs are used. (default: -1)
  --model model         Trained .json model (version 2.0+ models only)
                        (default: None)
  --motif motif         Score target motif declared using an extended dot-
                        bracket notation. Paired and unpaired bases are
                        denoted using parentheses '()' and dots '.',
                        respectively. A stretch of consecutive characters is
                        declared using the format <char>{<from>, <to>}. Can be
                        used in conjunction with --mask to modify the expected
                        underlying sequence of pairing states. (default: None)
  --path path           Target binary state sequence. When used in conjunction
                        with --motif, sequence constraintsfrom the motif
                        applied, but the state sequence provided by --path is
                        used to computeraw scores. (default: None)
  --hairpins            Score a representative set of hairpins (stem lengths 4
                        to 15; loop lengths 3 to 10). Automatically enabled
                        when the --HDSL flag is used. This flag overrides any
                        motif syntaxes provided via --motif. (default: False)
  --posteriors          Output the posterior probabilities of pairing states
                        (i.e. the probability Trellis) (default: False)
  --viterbi             Output the most likely sequence of pairing states for
                        entire transcripts (i.e. Viterbi paths) (default:
                        False)
  --HDSL                Use scores a representative set of hairpins (stem
                        lengths 4 to 15; loop lengths 3 to 10) to quantify
                        structuredness across the input data. This flag
                        overrides any motif syntaxes provided via --motif and
                        also activates --posteriors. (default: False)
  --SPP                 Smoothed P(paired). Quantifies structuredness across
                        the input data via local pairing probabilities. This
                        flag activates --posteriors. (default: False)
  --nan                 To attempt statistical inferences on the pairing state
                        of nucleotides with missing data when training, set
                        this flag. Note that this can lead to meaningless
                        results if observation quality is low or long runs of
                        missing data exist in the data. (default: False)
  --print-nan           Include NaN scores when writing scores to file. If the
                        data contain large runs ofmissing data, setting this
                        flag may make score files very large. (default: False)
  --no-prompt           Do not prompt a question if existing output files
                        could be overwritten. Files in output directory will
                        be overwritten if present. Useful for automation using
                        scripts or for running patteRNA on computing servers.
                        (default: False)
  --min-cscores min     Minimum number of scores to sample during construction
                        of null distributions to usefor c-score normalization
                        (default: 1000)
  --no-cscores          Suppress the computation of c-scores during the
                        scoring phase (default: False)
  --batch-size size     Number of transcripts to process at once using a pool
                        of parallel workers (default: 100)
  -c length, --context length
                        Flanking distance to use when computing motif energy
                        loss (default: 40)
```


### Inputs
patteRNA uses a FASTA-like convention for probing data (see this [example file](sample_data/weeks_set.shape)). As patteRNA learns from data, non-normalized data can be used directly. Also, patteRNA fully supports negatives and zero values, even when applying a log-transformation to the data (via the `-l` flag). We recommend to **not** artificially set negative values to 0. Missing data values must be set to `nan`, `NA` or `-999`.

### Training a model on a new dataset

By default, patteRNA learns its model from the data. Run an example training phase using the command:

```
patteRNA sample_data/weeks_set.shape sample_output -vl
```

> If you ran the test during installation, you will be prompted about overwriting files in the existing directory `test`. Answer `y`/`yes`. Note that in this example we run patteRNA in verbose-mode (`-v`) and we log transform (`-l`) the input data.

This command will generate an output folder `sample_output` in the current directory which contains:

- A log file: `<date>.log`
- Trained model: `trained_model.json`
- A plot of the fitted data: `fit.png`/`fit.svg`
- A plot of the model's log-likelihood convergence: `logL.png`/`logL.svg`

### Scoring
patteRNA supports structural motifs (via the `--motif` flag) that contain no gaps. These options can be used in conjunction with training to perform both training and scoring using a single command. However, we recommend to train patteRNA first and use the trained model in subsequent searches for motifs. Trained models are saved as `trained_model.json` and can be loaded using the flag `--model`.

#### Motif Specification
Standard motifs can be declared using an extended dot-bracket notation where stretches of consecutive repeats are denoted by curly brackets. For instance, an hairpin of stem size 4 and loop size 5 can be declared by `((((.....))))` (full form) or alternatively `({4}.{5}){4}` (short form). Curly brackets can also be used to indicate stretches of varying length using the convention `{<from>,<to>}`. For example, all loops of size 2 to 7 can be declared as `.{2,7}`. By default, RNA sequences are used to ensure a scored region sequence is compatible with the folding of the motif. RNA sequences must be provided in a FASTA file inputted using the option `-f <fasta-file>`. See [example commands](#examples).

#### Score Files
The results of a motif search are saved in the file `scores.txt` in the output directory. This file contains putative sites  the following columns:

- Transcript name
- Site start position (uses a 1-based encoding)
- Score
- c-score
- Binary cross entropy (BCE)
- Motif energy loss (MEL)
- Prob(motif) (computed via logistic binary classifier)
- Motif in dot-bracket notation
- Nucleotide sequence

### Additional outputs
#### Viterbi path
patteRNA can return the most likely sequence of pairing states across an entire transcript, called the Viterbi path, using the `--viterbi` flag. This will create a FASTA-like file called `viterbi.txt` in the output directory, with numerical pairing states encoded as 0/1 for unpaired/paired bases, respectively.

#### Posterior probabilities
The posterior probabilities of pairing states at each nucleotides can be requested using the flag `--posteriors`. This will output a FASTA-like file called `posteriors.txt` where the first and second lines (after the header) correspond to unpaired and paired probabilities, respectively.

#### Hairpin-derived structure level (HDSL)
HDSL is a measure of local structure that assists in converting patteRNA's predicted hairpins into a quantitative assenment of structuredness. This will output a FASTA-like file called `hdsl.txt` with HDSL profiles for all transcripts in the input data.

### Examples 

* Train the model and search for loops of length 5:

    ```
    patteRNA sample_data/weeks_set.shape example_outputs/loop -vl --motif ".{5}" -f sample_data/weeks_set.fa
    ```

* Search for all loops of length 5 using a trained model:

    ```
    patteRNA sample_data/weeks_set.shape example_outputs/loop_pretrained -vl --model test/trained_model.json --motif ".{5}" -f sample_data/weeks_set.fa
    ```

* Search for hairpins of variable stem size 4 to 6 and loop size 5:

    ```
    patteRNA sample_data/weeks_set.shape example_outputs/hairpin -vl --model test/trained_model.json -f sample_data/weeks_set.fa --motif "({4,6}.{5}){4,6}"
    ```


* Request HDSL profiles and the posterior state probabilities using a trained model:

    ```
    patteRNA sample_data/weeks_set.shape example_outputs/hdsl -vl --model test/trained_model.json --HDSL
    ```
  
* Train a model using a set of reference transcripts:

    ```
    patteRNA sample_data/weeks_set.shape example_outputs/loop -vl -f sample_data/weeks_set.fa --reference sample_data/weeks_set.dot
    ```
  
## Citation


CHANGES.md
# Changelog
All notable changes to patteRNA are documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.1] - 2021-10-20
### Changes
- Improved documentation [PR]
- Improved testing [PR]

### Fixed
- Improved logging when using reference structures [PR]


## [2.1-beta] - 2021-08-29
### Changes
- Stable implementation of logistic scoring classifier (can be turned off with `--no-vienna` or `--no-cscores` flags) [PR]
  - Added new dependency to `ViennaRNA` Python interface
- Added `--nan` flag to interpret missing data and `--print-nan` flag to print NaN scores [PR]
- Improved documentation [PR]

### Fixed
- Outputs for a transcript are computed and written all at once instead of on-the-fly during parallel processing, speeding up scoring [PR]


## [2.0] - 2021-08-24
### Changed
- Re-implemented `--reference` flag to natively work with new training pipeline [PR]
- Code optimization and cleanup

## [2.0.0-beta] - 2021-07-11
### Changed
- Fixed bug in scoring phase that rendered c-score compilation non-deterministic. [PR]
- Addition of `--SPP` flag to compute smoothed P(paired) as a structure profile. [PR]
- Improved documentation [PR]

## [2.0.0-alpha] - 2021-02-23
### Changed
- Major version release with almost a full rewrite of the method. [PR]
- Addition of `--HDSL` flag to compute local structure levels. [PR]
- Addition of a new Discretized Observation Model (DOM) emission model scheme, which is more precise for scoring and faster than a GMM. [PR]
- New c-score distribution sampling procedure is much more efficient than before. [PR]
- Now using human-readable `.json` format for saving a loading trained models. [PR]
- Matplotlib backend to `svg`. [PR]
- Dependencies: `humanfriendly`. [PR]

## [1.2.2] - 2020-01-08
### Changed
- Updated sample data with corrected structures. [PR]
- Updated README to reflect current developers. [PR]
- Fixed sample data. [PR]

## [1.2.1] - 2019-04-10
### Changed
- Installation procedure. [ML]
- Matplotlib backend to `Agg`. [ML]

## [1.2.0] - 2018-06-11
- Supervised initialization of Model's parameters based on reference RNA secondary structures in dot-bracket notation supplied via the new `--reference` flag. Note that `--reference` supports RNAstructure's `ct2dot` output format. [ML]
- Simulation framework for testing (devs only). [ML]
- Scoring motifs now returns, by default, a c-score based on a fitted Null distribution in addition to the original score. [PR/ML]
- Flag `--no-cscores` to turn off the computation of c-scores. [ML]
- The training set is now built automatically using KL divergence metrics (via option `--KL-div`). Data-dense transcripts are prioritized. [ML]
- Infinite values in structure profiles are now supported. [ML]
- Added a checkpoint to ensure paired/unpaired states are never flipped in the model. [ML]
- Dependencies: `cairosvg` (needed by `pygal`). [ML]
- Dependencies: `matplotlib`. [ML]

### Changed
- Motifs were all observations are missing now return NaN scores. [ML]
- Progress bar during scoring tracks individual transcripts instead of batches. [PR]
- The behavior of the `--min-density/-d` CLI flag was changed to affect both training and scoring. [ML]
- The default value for the `--min-density/-d` CLI flag was changed to 0 (i.e. all transcripts are used by default). [ML]
- Now renders all plots as PNG instead of SVG. [ML]
- Re-vamped user messages printed during the task. [ML]

### Removed
- Removed `-n` CLI flag (obsolete). [ML]
- Removed `--filter-test` CLI flag (obsolete). [ML]


## [1.1.4] - 2018-02-13
### Fixed
- Bugfix. Sequence constraints contained a bug affecting non fully nested target motifs. [PR]

## [1.1.3] - 2018-02-13
### Fixed
- Bugfix. Output Viterbi and posterior files were not deleted if already existing. [ML]


## [1.1.2] - 2018-02-06
### Changed
- Unsupervised initialization now uses by default an initial transition probability matrix derived from the Weeks set and GMM means based on data percentiles for increased robustness. [ML]

### Fixed
- Bugfix over v1.1.1 which was removed [ML]

## [1.1.0] - 2018-01-18
### Added
- The number of Gaussian components (`-k`) can be determined automatically using Aikaike Information Criteria (AICs). [ML]
- Automatic detection of the experimental assay based on the input observation filename extension. [ML]
- Motifs are sorted by scores in `scores.txt`. [ML]
- Dependencies: `scikit-learn` and `scipy` (needed by `scikit-learn`). [ML]
- Motif dot-brackets are declared using the option `--motif` and by default sequence constraints are applied. [ML]
- Pairing state sequences are declared using the option `--path` and used either alone or as a mask to `--motif`. [ML]

### Changed
- Structure profiling observations were removed from the output score file `scores.txt` to minimize file size. [ML]
- Versioned the latest tested patteRNA distributions to `latest`. [ML]
- Refactored `file_handler.py` to handle all FASTA-like files within a single function. [ML]
- Option `--gammas` is now named `--posteriors`. [ML]
- Option `--nogmm` is now devs only. [ML]
- Options `--pattern` and `-s/--seq` are now legacy and will be deprecated in the future. [ML]

### Removed
- Removed parameter for final state probabilities `rho` (obsolete). [ML]
- Removed `-wmin` as it is now obsolete. [ML]
- Removed `--PARS` CLI flag as file extensions are used to determine the assay (obsolete). [ML]
- Removed `--debug` CLI flag (obsolete). [ML]

### Fixed
- Fixed a bug where the last entry of input files would not be read. [ML]
- Minor bugfixes. [ML]
- Minor runtime optimizations. [ML]

## [1.0.0] - 2017-12-12
### Added
- Initial release. [ML]

src/patteRNA/arglib.py
"""Parse and handle input arguments"""

import argparse
import logging
import multiprocessing
from patteRNA import version

logger = logging.getLogger(__name__)


def parse_cl_args(inputargs):
    """
    Parse command line arguments.
    Args:
        inputargs (list): List of input arguments.

    Returns:
        input_file (dict): Dictionary of input file names
        run_config (dict): Dictionary of run configuration options

    """
    parser = argparse.ArgumentParser(prog="patteRNA",
                                     description="Rapid mining of RNA secondary structure motifs from profiling data.",
                                     epilog="",
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument("--version",
                        action="version",
                        version="%(prog)s {}".format(version.__version__))
    parser.add_argument("probing",
                        metavar="probing",
                        type=str,
                        help="FASTA-like file of probing data")
    parser.add_argument("output",
                        metavar="output",
                        type=str,
                        help="Output directory")
    parser.add_argument("-v", "--verbose",
                        action="store_true",
                        help="Print detailed progress logs")
    parser.add_argument("-f", "--fasta",
                        metavar="fasta",
                        default=None,
                        type=str,
                        help="FASTA file of RNA sequences")
    parser.add_argument("--reference",
                        metavar="reference",
                        default=None,
                        type=str,
                        help="FASTA-like file of reference RNA secondary structures in dot-bracket notation")
    parser.add_argument("-l", "--log",
                        action="store_true",
                        help="Log transform input data")
    parser.add_argument('--no-vienna',
                        action='store_true',
                        help="Do not attempt to use ViennaRNA libraries. Turns off LBC scoring classifier.")
    parser.add_argument("--GMM",
                        action="store_true",
                        help="Train a Gaussian Mixture Model (GMM) during training instead of a Discretized Observation"
                             "Model (DOM)")
    parser.add_argument("-k",
                        metavar="kernels",
                        type=int,
                        default=-1,
                        help="Number of kernels per pairing state to use in the emission model. By default, k is "
                             "determined automatically using Bayesian Information Criteria. Increasing k manually can "
                             "more precisely fit=the data, but could result in overfitting. Fitted data should always "
                             "be visually inspected after training to gauge if the model is adequate.")
    parser.add_argument("--KL-div",
                        metavar="KL-div",
                        type=float,
                        default=0.001,
                        help="Minimum Kullback–Leibler divergence criterion for building the training set. The KL "
                             "divergence measures the difference in information content between the full dataset "
                             "and the training set. The smaller the value, the more representative the training "
                             "set will be with respect to the full dataset. However, this will produce a "
                             "larger training set and increase both runtime and RAM consumption during training.")
    parser.add_argument("-e", "--epsilon",
                        metavar="eps",
                        type=float,
                        default=1e-2,
                        help="Convergence criterion")
    parser.add_argument("-i", "--maxiter",
                        metavar="iter",
                        type=int,
                        default=250,
                        help="Maximum number of training iterations")
    parser.add_argument("-t", "--n-tasks",
                        metavar="tasks",
                        type=int,
                        default=-1,
                        help="Number of parallel processes. By default all available CPUs are used.")
    parser.add_argument("--model",
                        metavar="model",
                        type=str,
                        default=None,
                        help="Trained .json model (version 2.0+ models only)")
    parser.add_argument("--motif",
                        metavar="motif",
                        type=str,
                        default=None,
                        help="Score target motif declared using an extended dot-bracket notation. Paired and unpaired "
                             "bases are denoted using parentheses '()' and dots '.', respectively. A stretch of "
                             "consecutive characters is declared using the format <char>{<from>, <to>}. Can be used in "
                             "conjunction with --mask to modify the expected underlying sequence of pairing states.")
    parser.add_argument("--path",
                        metavar="path",
                        type=str,
                        default=None,
                        help="Target binary state sequence. When used in conjunction with --motif, sequence constraints"
                             "from the motif applied, but the state sequence provided by --path is used to compute"
                             "raw scores.")
    parser.add_argument("--hairpins",
                        action="store_true",
                        help="Score a representative set of hairpins (stem lengths 4 to 15; loop lengths 3 to 10). "
                             "Automatically enabled when the --HDSL flag is used. This flag overrides any motif "
                             "syntaxes provided via --motif.")
    parser.add_argument("--posteriors",
                        action="store_true",
                        help="Output the posterior probabilities of pairing states (i.e. the probability Trellis)")
    parser.add_argument("--viterbi",
                        action="store_true",
                        help="Output the most likely sequence of pairing states for entire transcripts (i.e. Viterbi "
                             "paths)")
    parser.add_argument("--HDSL",
                        action="store_true",
                        help="Use scores a representative set of hairpins (stem lengths 4 to 15; loop lengths 3 to 10) "
                             "to quantify structuredness across the input data. This flag overrides any motif "
                             "syntaxes provided via --motif and also activates --posteriors.")
    parser.add_argument("--SPP",
                        action="store_true",
                        help="Smoothed P(paired). Quantifies structuredness across the input data via local pairing "
                             "probabilities. This flag activates --posteriors.")
    parser.add_argument("--nan",
                        action="store_true",
                        help="To attempt statistical inferences on the pairing state of nucleotides with missing "
                             "data when training, set this flag. Note that this can lead to meaningless results if "
                             "observation quality is low or long runs of missing data exist in the data.")
    parser.add_argument("--print-nan",
                        action="store_true",
                        help="Include NaN scores when writing scores to file. If the data contain large runs of"
                             "missing data, setting this flag may make score files very large.")
    parser.add_argument("--no-prompt",
                        action="store_true",
                        help="Do not prompt a question if existing output files could be overwritten. Files in output "
                             "directory will be overwritten if present. Useful for automation using scripts or for "
                             "running patteRNA on computing servers. ")
    parser.add_argument("--min-cscores",
                        metavar="min",
                        type=int,
                        default=1000,
                        help="Minimum number of scores to sample during construction of null distributions to use"
                             "for c-score normalization")
    parser.add_argument("--no-cscores",
                        action="store_true",
                        help="Suppress the computation of c-scores during the scoring phase")
    parser.add_argument("--batch-size",
                        metavar="size",
                        type=int,
                        default=100,
                        help="Number of transcripts to process at once using a pool of parallel workers")
    parser.add_argument('-c', '--context',
                        metavar="length",
                        type=int,
                        default=40,
                        help="Flanking distance to use when computing motif energy loss")

    args = parser.parse_args(inputargs)  # Parse input args

    input_files = dict.fromkeys(['probing', 'fasta', 'reference', 'model'], None)
    run_config = dict()

    for attr, value in args.__dict__.items():

        if attr in input_files.keys():
            input_files[attr] = value
        else:
            run_config[attr] = value

    # Process HDSL constraints
    if run_config['HDSL']:
        if run_config['no_cscores']:
            logger.error("Configuration error: --HDSL cannot be used in conjunction with --no-cscores")
        if run_config['motif'] is not None:
            logger.warning('Configuration warning: --motif cannot be used in conjunction with --HDSL')

    # Process hairpin constraints
    if run_config['hairpins']:
        if run_config['motif'] is not None:
            logger.warning('Configuration warning: --motif cannot be used in conjunction with --hairpins')

    if run_config['HDSL']:
        run_config['motif'] = "({4,15}.{3,10}){4,15}"
        run_config['hairpins'] = True
        run_config['posteriors'] = True

    if run_config['hairpins']:
        run_config['motif'] = "({4,15}.{3,10}){4,15}"

    if run_config['SPP']:
        run_config['posteriors'] = True

    # Set training configuration flag based on provided model
    if input_files['model'] is not None:
        run_config['training'] = False
    else:
        run_config['training'] = True

    # Set scoring configuration flag
    if run_config['posteriors'] or run_config['motif'] or run_config['viterbi']:
        run_config['scoring'] = True
    else:
        run_config['scoring'] = False

    if run_config['n_tasks'] == -1:
        run_config['n_tasks'] = multiprocessing.cpu_count()

    run_config['reference'] = False  # Even if reference data is provided, we need to inspect it before deciding to
    # actually use it for training. So, for now, run_config['reference'] is False.
    run_config['hdsl_params'] = (1.2, 0.5)

    return input_files, run_config


def summarize_job(input_files, run_config):
    """
    Generate concise text summary of config arguments for command line printing.

    Args:
        input_files (dict): Dictionary of input files
        run_config (dict): Dictionary of configuration parameters

    Returns:
        text (str): String representation of run configuration
    """

    space = '\n'
    tab = ' ' * 7
    text = "Summarizing job ... \n"
    text += space
    text += tab + "Input files:\n"
    text += 2 * tab + "probing: {}\n".format(input_files['probing'])
    if input_files['fasta'] is not None:
        text += 2 * tab + "fasta: {}\n".format(input_files['fasta'])
    if input_files['reference'] is not None:
        text += 2 * tab + "reference: {}\n".format(input_files['reference'])
    if input_files['model'] is not None:
        text += 2 * tab + "model: {}\n".format(input_files['model'])
    text += space

    text += tab + "Configuration parameters:\n"
    text += 2 * tab + "log: {}\n".format(run_config['log'])
    text += 2 * tab + "training: {}\n".format(run_config['training'])
    text += 2 * tab + "scoring: {}\n".format(run_config['scoring'])
    if run_config['k'] == -1:
        text += 2 * tab + "k: auto\n"
    else:
        text += 2 * tab + "k: {}\n".format(run_config['k'])
    if run_config['GMM']:
        text += 2 * tab + "GMM: True\n"
    else:
        text += 2 * tab + "DOM: True\n"
    if run_config['hairpins']:
        text += 2 * tab + "hairpins: {}\n".format(run_config['hairpins'])
    if run_config['motif'] is not None:
        text += 2 * tab + "motif: {}\n".format(run_config['motif'])
    if run_config['path'] is not None:
        text += 2 * tab + "path: {}\n".format(run_config['path'])
    if run_config['posteriors']:
        text += 2 * tab + "posteriors: {}\n".format(run_config['posteriors'])
    if run_config['HDSL']:
        text += 2 * tab + "HDSL: {}\n".format(run_config['HDSL'])
    if run_config['SPP']:
        text += 2 * tab + "SPP: {}\n".format(run_config['SPP'])
    if run_config['viterbi']:
        text += 2 * tab + "viterbi: {}\n".format(run_config['viterbi'])
    text += space

    text += tab + "Using {} parallel processes\n".format(run_config['n_tasks'])

    return text


def summarize_config(input_files, run_config):
    """
    Generate exhaustive text summary of config arguments for logger.

    Args:
        input_files (dict): Dictionary of input files
        run_config (dict): Dictionary of configuration parameters

    Returns:
        text (str): String representation of run configuration
    """

    hline = "========================================================="
    text = "Summarizing configuration:" \
           "\n{}\n" \
           "Running patteRNA with the following parameters:" \
           "\n{}\n".format(hline, hline)

    text += "Input files:\n"
    for key in input_files:
        if input_files[key]:
            text += "\t{}: {}\n".format(key, input_files[key])

    text += "\nConfiguration parameters:\n"
    for key in run_config:
        text += "\t{}: {}\n".format(key, run_config[key])

    text += hline

    return text


if __name__ == '__main__':
    pass
